File: agent-governance-typescript/src/context-audit.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.
/**
 * Audit events for context envelope transitions.
 *
 * Emits CONTEXT_* events describing how an envelope changed across an action
 * or delegation. Every event carries its own classification floor — at least
 * the envelope's aggregate sensitivity.
 *
 * Parity with agent-governance-python agent-os/src/agent_os/policies/context_audit.py.
 * Refs #3084.
 */

import { DataClassification } from './data-classification';
import type { ContextEnvelope } from './context-envelope';

export const CONTEXT_ENVELOPE_CREATED = 'CONTEXT_ENVELOPE_CREATED';
export const CONTEXT_ENVELOPE_UPDATED = 'CONTEXT_ENVELOPE_UPDATED';
export const CONTEXT_AGGREGATION_ELEVATED = 'CONTEXT_AGGREGATION_ELEVATED';
export const CONTEXT_DELEGATED = 'CONTEXT_DELEGATED';
export const CONTEXT_REDACTED = 'CONTEXT_REDACTED';
export const DERIVED_ARTIFACT_LABELED = 'DERIVED_ARTIFACT_LABELED';

/** A recorded transition between two envelope versions. */
export interface ContextEvent {
  readonly eventType: string;
  readonly agentId: string;
  readonly contextEnvelopeId: string;
  readonly previousSensitivity: DataClassification;
  readonly newSensitivity: DataClassification;
  readonly labelsAdded: ReadonlySet<string>;
  readonly rulesApplied: readonly string[];
  readonly restrictionsAdded: ReadonlySet<string>;
  readonly classification: DataClassification;
}

/** Build a ContextEvent describing the transition before -> after. */
export function contextEvent(
  eventType: string,
  agentId: string,
  before: ContextEnvelope,
  after: ContextEnvelope,
  rulesApplied: readonly string[] = [],
): ContextEvent {
  const classification = Math.max(
    before.aggregateSensitivity,
    after.aggregateSensitivity,
  ) as DataClassification;

  const labelsAdded = new Set([...after.labels].filter((l) => !before.labels.has(l)));
  const restrictionsAdded = new Set(
    [...after.restrictions].filter((r) => !before.restrictions.has(r)),
  );

  return {
    eventType,
    agentId,
    contextEnvelopeId: after.envelopeId,
    previousSensitivity: before.aggregateSensitivity,
    newSensitivity: after.aggregateSensitivity,
    labelsAdded,
    rulesApplied,
    restrictionsAdded,
    classification,
  };
}

File: agent-governance-golang/packages/agentmesh/context_audit.go
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

package agentmesh

import "strings"

// ContextEventType is the stable wire type for context-governance audit events.
type ContextEventType string

const (
	// ContextEnvelopeCreated records creation of a context envelope.
	ContextEnvelopeCreated ContextEventType = "CONTEXT_ENVELOPE_CREATED"
	// ContextEnvelopeUpdated records a context envelope transition.
	ContextEnvelopeUpdated ContextEventType = "CONTEXT_ENVELOPE_UPDATED"
	// ContextAggregationElevated records an aggregation-driven sensitivity increase.
	ContextAggregationElevated ContextEventType = "CONTEXT_AGGREGATION_ELEVATED"
	// ContextDelegated records propagation of context into a child delegation.
	ContextDelegated ContextEventType = "CONTEXT_DELEGATED"
	// ContextRedacted records a redaction applied to governed context.
	ContextRedacted ContextEventType = "CONTEXT_REDACTED"
	// DerivedArtifactLabeled records classification of an artifact derived from context.
	DerivedArtifactLabeled ContextEventType = "DERIVED_ARTIFACT_LABELED"
)

// ContextEvent records the governance-relevant delta between two envelope versions.
// Its Classification is never lower than either envelope's aggregate sensitivity.
type ContextEvent struct {
	EventType           ContextEventType   `json:"event_type" yaml:"event_type"`
	AgentID             string             `json:"agent_id" yaml:"agent_id"`
	ContextEnvelopeID   string             `json:"context_envelope_id" yaml:"context_envelope_id"`
	PreviousSensitivity DataClassification `json:"previous_sensitivity" yaml:"previous_sensitivity"`
	NewSensitivity      DataClassification `json:"new_sensitivity" yaml:"new_sensitivity"`
	LabelsAdded         []string           `json:"labels_added" yaml:"labels_added"`
	RulesApplied        []string           `json:"rules_applied" yaml:"rules_applied"`
	RestrictionsAdded   []string           `json:"restrictions_added" yaml:"restrictions_added"`
	Classification      DataClassification `json:"classification" yaml:"classification"`
}

// ContextEventFor builds a deterministic event describing before -> after.
func ContextEventFor(
	eventType ContextEventType,
	agentID string,
	before ContextEnvelope,
	after ContextEnvelope,
	rulesApplied []string,
) ContextEvent {
	before = cloneContextEnvelope(before)
	after = cloneContextEnvelope(after)

	classification := before.AggregateSensitivity
	if after.AggregateSensitivity > classification {
		classification = after.AggregateSensitivity
	}

	return ContextEvent{
		EventType:           eventType,
		AgentID:             agentID,
		ContextEnvelopeID:   after.EnvelopeID,
		PreviousSensitivity: before.AggregateSensitivity,
		NewSensitivity:      after.AggregateSensitivity,
		LabelsAdded:         addedTokens(before.Labels, after.Labels),
		RulesApplied:        cloneTokens(rulesApplied),
		RestrictionsAdded:   addedTokens(before.Restrictions, after.Restrictions),
		Classification:      classification,
	}
}

func addedTokens(before []string, after []string) []string {
	beforeSet := tokenSet(before)
	added := make(map[string]bool)
	for _, token := range after {
		normalized := strings.TrimSpace(token)
		if normalized != "" && !beforeSet[normalized] {
			added[normalized] = true
		}
	}
	return cloneTokens(sortedTokens(added))
}
